python-package/lightgbm/sklearn.py

```diff
@@ -1087,16 +1087,15 @@ def fit(  # type: ignore[override]
             eval_metric = eval_metric_list
 
         # do not modify args, as it causes errors in model selection tools
-        valid_sets = None
+        valid_sets: List[Tuple] = []
```

Comment:
Do you see a user-facing way that it changes behavior?

I don't see one.

`valid_sets` in `LGBMClassifier.fit()` is passed through to `LGBMModel.fit()` via argument `eval_set`.

https://github.com/microsoft/LightGBM/blob/fe15761b683e796cf68d097b301d171bbffd6aff/python-package/lightgbm/sklearn.py#L1106

`LGBMModel.fit()` initializes an internal empty list, `valid_sets`.

https://github.com/microsoft/LightGBM/blob/fe15761b683e796cf68d097b301d171bbffd6aff/python-package/lightgbm/sklearn.py#L758

If the passed-in `eval_set` not `None`, you're right that *technically* another code path is considered

https://github.com/microsoft/LightGBM/blob/fe15761b683e796cf68d097b301d171bbffd6aff/python-package/lightgbm/sklearn.py#L759

However, that won't actually do anything differently in the case `eval_set = []` vs. `eval_set = None`, because all the code inside that path is a `for` loop over `eval_set` (and a for loop over an empty list is basically the same as the code inside the `for loop` being skipped entirely).

https://github.com/microsoft/LightGBM/blob/fe15761b683e796cf68d097b301d171bbffd6aff/python-package/lightgbm/sklearn.py#L773

And after that, whether `eval_set = []` or `eval_set = None`, the `valid_sets` internal variable will still be `[]`, and that's what gets passed through to `lgb.train()`.

https://github.com/microsoft/LightGBM/blob/fe15761b683e796cf68d097b301d171bbffd6aff/python-package/lightgbm/sklearn.py#L811

So unless you see another path, I think the change proposed in this PR is safe and won't be visible to users at all.

Reply:
Thanks for pointing this out by the way! I won't pretend that I did this super-careful analysis before proposing this change. I was just relying on having good test coverage of this part of the code + the fact that `lgb.train()` accepts both `None` and a list.

I also am getting confused by the switching notation back and forth between `eval_set` and `valid_sets` to describe the same thing, so definitely double-check that I've understood this correctly.

Reply:
Would you consider doing something like this instead?
```python
valid_sets: Optional[List[Tuple]] = None
if eval_set is not None:
    valid_sets = []
```
that also removes the warnings and doesn't change the current behavior.

Reply:
That'll fix one error, but not this one:

```text
python-package/lightgbm/sklearn.py:1097: error: List item 0 has incompatible type "None"; expected "Tuple[Any, ...]"  [list-item]
```

Because of this line:

https://github.com/microsoft/LightGBM/blob/90a4510c78ecae63c087e076383e57ebfdfc9796/python-package/lightgbm/sklearn.py#L1095

I tried then changing to something like this.

```python
valid_sets: Optional[List[Tuple]] = None
if eval_set is not None:
    valid_sets = []
if eval_set is not None:
    if isinstance(eval_set, tuple):
        eval_set = [eval_set]
    for i, (valid_x, valid_y) in enumerate(eval_set):
        if valid_x is X and valid_y is y:
            valid_sets.append((valid_x, _y))
        else:
            valid_sets[i].append((valid_x, self._le.transform(valid_y)))
```

But that doesn't work either, because the `Optional` implies that `valid_sets` can be `None`.

```text
python-package/lightgbm/sklearn.py:1065: error: Argument 1 to "len" has incompatible type "Optional[ndarray[Any, Any]]"; expected "Sized"  [arg-type]
python-package/lightgbm/sklearn.py:1099: error: Item "None" of "Optional[List[Tuple[Any, ...]]]" has no attribute "append"  [union-attr]
python-package/lightgbm/sklearn.py:1101: error: Value of type "Optional[List[Tuple[Any, ...]]]" is not indexable  [index]
python-package/lightgbm/sklearn.py:1101: error: Item "Tuple[Any, ...]" of "Union[Tuple[Any, ...], Any]" has no attribute "append"  [union-attr]
```

>  and doesn't change the current behavior

Can you point to any specific user-facing way that the behavior changes?

I understand that `None` is different from `[]` but based on my investigation in https://github.com/microsoft/LightGBM/pull/5724#discussion_r1117986421 I don't see any actual difference in what `lightgbm` does with `valid_sets=None` vs. `valid_sets=[]`.

Reply:
I mean this diff (wrt master):
```diff
diff --git a/python-package/lightgbm/sklearn.py b/python-package/lightgbm/sklearn.py
index 3a246c77..cc0bd84b 100644
--- a/python-package/lightgbm/sklearn.py
+++ b/python-package/lightgbm/sklearn.py
@@ -1088,16 +1088,16 @@ class LGBMClassifier(_LGBMClassifierBase, LGBMModel):
             eval_metric = eval_metric_list

         # do not modify args, as it causes errors in model selection tools
-        valid_sets = None
+        valid_sets: Optional[List[Tuple]] = None
         if eval_set is not None:
             if isinstance(eval_set, tuple):
                 eval_set = [eval_set]
-            valid_sets = [None] * len(eval_set)
-            for i, (valid_x, valid_y) in enumerate(eval_set):
+            valid_sets = []
+            for valid_x, valid_y in eval_set:
                 if valid_x is X and valid_y is y:
-                    valid_sets[i] = (valid_x, _y)
+                    valid_sets.append((valid_x, _y))
                 else:
-                    valid_sets[i] = (valid_x, self._le.transform(valid_y))
+                    valid_sets.append((valid_x, self._le.transform(valid_y)))

         super().fit(
             X,
```
raises only these warnings (mypy 1.1.1):
```
python-package/lightgbm/sklearn.py:1065: error: Argument 1 to "len" has incompatible type "Optional[ndarray[Any, Any]]"; expected "Sized"  [arg-type]
python-package/lightgbm/sklearn.py:1183: error: Unsupported operand types for < ("int" and "None")  [operator]
python-package/lightgbm/sklearn.py:1183: note: Left operand is of type "Optional[int]"
python-package/lightgbm/sklearn.py:1202: error: Incompatible return value type (got "Optional[ndarray[Any, Any]]", expected "ndarray[Any, Any]")  [return-value]
python-package/lightgbm/sklearn.py:1209: error: Incompatible return value type (got "Optional[int]", expected "int")  [return-value]
```
which don't seem to be related to the valid sets.

Reply:
OOHHHHHHHHHHHHHHHHHHHH 

I see. Thanks for taking the time to share the diff like that, I totally understand now. You're right, that's cleaner and eliminates the errors this PR was attempting to eliminate.

Just pushed https://github.com/microsoft/LightGBM/pull/5724/commits/212962489e59a3c67a1fba8a780339fc04088171 switching to that.

Reply:
Thanks! Sorry for not being clear enough before.

Reply:
All good, thanks as always for your thorough review!